### PR TITLE
feat: display membership overlay on posts route for PRO posts

### PR DESCRIPTION
**Pull request by zacjones93**
We are setting access in course builder but egghead is displaying all lessons regardless of access (`free` or `pro`). This introduces a membership overlay which prompts people to become a member or sign in

## Pro post [preview link](https://egghead-io-nextjs-git-zj-posts-go-pro-overlay-eggheadio1.vercel.app/karabiner-rethinking-right-click-as-a-macro-command-center~targ7)
(my stripe prices are messed up locally)
PRO post

## free post [preview link](https://egghead-io-nextjs-git-zj-posts-go-pro-overlay-eggheadio1.vercel.app/get-started-with-svelte-5-using-the-sv-cli~60uep)
free post

gif

### CODE REVIEW

File: src/components/pages/lessons/overlay/confirm-membership.tsx [hunk 1]
```diff
@@ -16,7 +16,9 @@ type HeaderProps = {
 type ConfirmMembershipProps = {
   sessionId: string
   viewLesson: Function
-  lesson: LessonResource
+  lesson: {
+    slug: string
+  }
```
Comment (zacjones93) [inline src/components/pages/lessons/overlay/confirm-membership.tsx:21]: These membership components only needed the lesson slug (or collection when in a course) so narrowed the type here to satisfy posts page requirements

File: src/components/pages/lessons/overlay/go-pro-cta-overlay.tsx [hunk 2]
```diff
@@ -22,9 +22,15 @@ import ConfirmMembership from './confirm-membership'
...
+import {usePathname} from 'next/navigation'
 
 type JoinCTAProps = {
-  lesson: LessonResource
+  lesson: {
+    slug: string
+    collection?: {
+      title: string
+    }
+  }
```
Comment (zacjones93) [inline src/components/pages/lessons/overlay/go-pro-cta-overlay.tsx:33]: `GoProCtaOverlay` only accesses lesson slug and `collection.title`

File: src/pages/[post].tsx [hunk 3]
```diff
@@ -26,6 +26,39 @@ import CopyToClipboard from '@/components/copy-resource'
...
+  createdAt: z.date(),
+  updatedAt: z.date(),
+  deletedAt: z.date().nullish(),
+  currentVersionId: z.string().optional(),
+  organizationId: z.string().nullish(),
+  createdByOrganizationMembershipId: z.string().nullish(),
+  name: z.string().optional(),
+  image: z.string().optional(),
+})
+export type Post = z.infer<typeof PostSchema>
```
Comment (zacjones93) [inline src/pages/[post].tsx:61]: created a post schema

File: src/pages/[post].tsx [hunk 4]
```diff
@@ -74,49 +107,112 @@ SELECT *
...
+}
+
+export const getStaticProps: GetServerSideProps = async function ({params}) {
+  if (!params?.post) {
+    return {
+      notFound: true,
+    }
+  }
+
+  const result = await getPost(params.post as string)
```
Comment (zacjones93) [inline src/pages/[post].tsx:207]: moved the getPost SQL into its own function

File: src/pages/[post].tsx [hunk 5]
```diff
@@ -265,7 +351,7 @@ export default function PostPage({
   tags,
 }: {
   mdxSource: any
-  post: any
+  post: Post
```
Comment (zacjones93) [inline src/pages/[post].tsx:354]: typing our posts now

File: src/pages/[post].tsx [hunk 6]
```diff
@@ -447,6 +536,12 @@ function PostPlayer({
   const {mutateAsync: addProgressToLesson} =
     trpc.progress.addProgressToLesson.useMutation()
 
+  const {data: viewer} = trpc.user.current.useQuery()
+
+  const isPro = post.fields.access === 'pro'
+  const canView =
+    !isPro || (isPro && Boolean(viewer) && Boolean(viewer?.is_pro))
```
Comment (zacjones93) [inline src/pages/[post].tsx:543]: canView logic

`(isPro && Boolean(viewer) && Boolean(viewer?.is_pro)` -> if the lesson has `pro` access level, `viewer` needs to be defined and viewer `is_pro` attribute set to true

File: src/pages/[post].tsx [hunk 7]
```diff
@@ -465,28 +560,34 @@ function PostPlayer({
...
+              currentTime: muxPlayer.currentTime,
+              lessonId: eggheadLessonId,
+            })
+            setWritingProgress(false)
+          }}
+          className="relative z-10 flex items-center max-h-[calc(100vh-240px)] h-full bg-black justify-center"
+        />
+      ) : (
+        <GoProCtaOverlay lesson={{slug: post.fields.slug ?? ''}} />
+      )}
```
Comment (zacjones93) [inline src/pages/[post].tsx:589]: display `GoProCtaOverlay` when `canView` is false

**Review by joelhooks — APPROVED** [review]

**Pull request merged**